File: 1725-number-of-sets-of-k-non-overlapping-line-segments/number-of-sets-of-k-non-overlapping-line-segments.py
class Solution:
    def numberOfSets(self, n: int, k: int) -> int:
        MOD = 10**9 + 7

        # dp[i][j] = number of ways to draw j segments
        # using points 0 to i
        dp = [[0] * (k + 1) for _ in range(n)]

        # 0 segments -> 1 way
        for i in range(n):
            dp[i][0] = 1

        for j in range(1, k + 1):
            prefix = 0

            for i in range(1, n):
                # Add ways from dp[i-1][j-1]
                prefix = (prefix + dp[i - 1][j - 1]) % MOD

                # Either don't use point i,
                # or finish a segment at point i
                dp[i][j] = (dp[i - 1][j] + prefix) % MOD

        return dp[n - 1][k]

File: 1725-number-of-sets-of-k-non-overlapping-line-segments/number-of-sets-of-k-non-overlapping-line-segments.java
class Solution {
    public int numberOfSets(int n, int k) {
        long MOD = 1000000007;

        // dp[i][j] = number of ways to draw j segments
        // using points 0 to i
        long[][] dp = new long[n][k + 1];

        // 0 segments can always be made in 1 way
        for (int i = 0; i < n; i++) {
            dp[i][0] = 1;
        }

        for (int j = 1; j <= k; j++) {

            long prefix = 0;

            for (int i = 1; i < n; i++) {

                // Add ways from dp[i-1][j-1]
                prefix = (prefix + dp[i - 1][j - 1]) % MOD;

                // Don't use point i OR finish a segment at i
                dp[i][j] = (dp[i - 1][j] + prefix) % MOD;
            }
        }

        return (int) dp[n - 1][k];
    }
}